### DISCUSSION TITLE
v1.2.0-preview2

### DISCUSSION THREAD

**Discussion by gabbybilka** [microsoft/WindowsAppSDK#3018] [Announcements]
## 1.2 Preview 2 (1.2.220930.4-preview2) 🔃

The second preview of WinAppSDK 1.2 is now available. This preview includes new features and bug fixes, including: 
- Third-party Widgets in Windows
- DisplayInformation APIs for HDR application support
- IL trimming for .NET applications to reduce your apps' disk footprint size

To see everything that's new and changed, check out the full [release notes](https://aka.ms/windowsappsdk/1.2/1.2-preview2/release-notes). 

The WinUI Gallery is also available with a 1.2-preview2 [branch](https://github.com/microsoft/WinUI-Gallery/tree/feature/winappsdk-1.2).

### Updating to this version
The 1.2-preview2 package is available at: https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.2.220930.4-preview2

#### Notes when updating to 1.2-preview2
- When you reference WinAppSDK 1.2 from a project you might see an error similar to: *"Detected package downgrade: Microsoft.Windows.SDK.BuildTools from 10.0.22621.1 to 10.0.22000.194."* To resolve, you can update the project’s reference to a more recent and compatible version of Microsoft.Windows.SDK.BuildTools, or simply remove the reference from your project to implicitly reference a compatible version.
- Starting with WinAppSDK 1.2-preview2, .NET 5 is no longer supported and C# apps will require a minimum version of .NET 6.

### Providing feedback
As we introduce new & upcoming features in our preview releases, we appreciate all of your feedback as we work towards a stable release:
- For WinUI 3 feedback, please open an issue on the [WinUI repo](https://github.com/microsoft/microsoft-ui-xaml/issues/new/choose).
- For feedback regarding all other Windows App SDK components, please open an issue here, on the [Windows App SDK repo](https://github.com/microsoft/WindowsAppSDK/issues/new/choose).

<hr /><em>This discussion was created from the release <a href='https://github.com/microsoft/WindowsAppSDK/releases/tag/v1.2.0-preview2'>v1.2.0-preview2</a>.</em>

**Comment by kmgallahan** [comment]
Using VS Version 17.4.0 Preview 2.1 with WASDK 1.2.0-preview2 I'm getting two different sets of build errors at different times after changing code. 

They go away if I rebuild the entire solution, but this is quite slow. They also weren't present before upgrading both today, although I'm not sure which upgrade is responsible.

```
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\Microsoft.Build.Msix.Packaging.targets(920,5): error APPX0002: Task 'WinAppSdkGenerateAppxManifest' failed. Could not load file or assembly 'Microsoft.Cci, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\Microsoft.Build.Msix.Packaging.targets(920,5): error APPX0002:
```

and

```
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018: The "ExpandPriContent" task failed unexpectedly.
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018: System.ObjectDisposedException: Cannot access a disposed object.
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018: Object name: 'it is not allowed to use disposed Microsoft.VisualStudio.Telemetry.TelemetrySession object.'.
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.VisualStudio.Telemetry.TelemetryDisposableObject.RequiresNotDisposed()
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.VisualStudio.Telemetry.TelemetrySession.SerializeSettings()
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.Build.AppxPackage.Telemetry.Get(IBuildEngine4 buildEngine, String vsTelemetrySession)
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.Build.AppxPackage.AppxPackagingTaskHelper..ctor(IBuildEngine4 buildEngine, TaskLoggingHelper logger, Action execute, Func`2 exceptionHandler, Action cleanup, Func`1 manifestFileNameGetter, String vsTelemetrySession)
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.Build.AppxPackage.AppxPackagingToolTaskBase.Execute()
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
2>C:\Users\kmgal\.nuget\packages\microsoft.windowsappsdk\1.2.220930.4-preview2\buildTransitive\MrtCore.PriGen.targets(911,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
```

**Reply by MikeHillberg** [reply]
After the full rebuild can you do incremental again? Or do you need to full rebuild every time you make a change?

**Reply by kmgallahan** [reply]
@MikeHillberg After some testing it appears the first 2 line error appears after any change to any .cs or .xaml file, requiring a full rebuild every time. 

The second change seems to only come up after a debugging session has already started and ended, and may be related to an internal VS resource not being released by the previous debugging session. Executing 'Start Debugging' again without rebuilding lets the debugging session begin without stopping for the error again, although I'm not sure if the most recent code changes are being used or not.

The first error isn't dependent on a debugging session having been previously started (since the last VS/machine reboot).

**Reply by kmgallahan** [reply]
@MikeHillberg Through the magic of time I'm no longer experiencing this problem.

**Reply by MikeHillberg** [reply]
Frustrating. Thanks for the update.

**Comment by sudongg** [comment]
Packaging error:
CS0006: Metadata file 'C:\App1\bin\x64\Release\net6.0-windows10.0.19041.0\win10-x64\test.dll' could not be found

If the project references a library project, and the library project contains Microsoft.WindowsAppSDK 1.2.220930.4-preview2 , it will fail to publish with this error.

The reason is that the dll location generated by the referenced project is 'C:\App1\bin\x64\Release\net6.0-windows10.0.19041.0\test.dll' instead of 'C:\App1\bin\x64\Release\net6 .0-windows10.0.19041.0\win10-x64\test.dll', it is trying to find the file in the wrong location, I manually copied the file to the specified location and packaged successfully.

**Reply by gabbybilka** [reply]
@DrusTheAxe @MikeHillberg FYI

**Reply by MikeHillberg** [reply]
Thanks for reporting this @sudongg. What do you mean by "publish"? When I have an app reference a library project I'm not seeing this when I F5. Could you open this as an Issue?

**Reply by sudongg** [reply]
@MikeHillberg 
Issue link:
https://github.com/microsoft/WindowsAppSDK/issues/3065
